### ISSUE TITLE
Invalid `type-mismatch` error with some `warp` filters

### ISSUE THREAD

**Issue by awaken1ng** [rust-analyzer/rust-analyzer#11798]
```toml
[dependencies]
warp = "^0.3"
```

```rs
use warp::{Filter, Rejection};

fn route() -> impl Filter<Extract = (String,), Error = Rejection> + Clone {
    warp::any()
        .and_then(|| async move {
            if false {
                return Err(warp::reject::not_found())
            }

            Ok(String::new())
        })
}

fn main() {}
```

RA raises invalid `type-mismatch` error:
```
expected Result<{unknown}, Rejection>, found impl Future<Output = Result<String, {unknown}>>
```

Code `cargo check`s and compiles fine otherwise, bisected to ab3313b1cb5e9ff79ecef0fb188873c892c193f1 (#11780)

**rustc version**: 1.59.0 (9d1b2106e 2022-02-23)

**Comment by saiintbrisson** [comment]
Can confirm on r-a `c2ea37892 2022-03-23 nightly`. Raising tons of erros on all `#[async_trait]` usages and async functions with `tracing`'s `instrument` attribute. 
![image](https://user-images.githubusercontent.com/29989290/159729806-a47a20ac-d885-4045-aa81-fa888cbe0d46.png)

**Comment by ErwanDL** [comment]
I've also started observing very weird type-mismatch errors a few hours ago (on code that wasn't raising any error before), notably with tokio tasks that contain infinite `loop`s: 

```
expected JoinHandle<()>, found JoinHandle<!> rust-analyzer(type-mismatch)
```

Also errors with `async_trait` usage:
```
expected Pin<Box<dyn Future<Output = Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError>> + Send, Global>>, found Pin<Box<impl Future<Output = Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError>>, Global>> rust-analyzer(type-mismatch)
expected Pin<Box<dyn Future<Output = Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError>> + Send, Global>>, found Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError> rust-analyzer(type-mismatch)
expected Pin<Box<dyn Future<Output = Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError>> + Send, Global>>, found Pin<Box<impl Future<Output = Result<HashSet<SignalingEndpoint, RandomState>, EndpointsError>>, Global>> rust-analyzer(type-mismatch)
```

**Comment by saiintbrisson** [comment]
> A more minimal reproduction (that doesn't involve all the warp machinery) would also be helpful.

```rust
#[async_trait::async_trait]
trait Foo {
    async fn bar();
}

struct Baz;

#[async_trait::async_trait]
impl Foo for Baz {
    async fn bar() {}
}
```
Raises `expected Pin<Box<dyn Future<Output = ()> + Send, Global>>, found Pin<Box<impl Future<Output = ()>, Global>>`.

**Comment by flodiebold** [comment]
Hm, that might actually just be a problem with our handling of `async fn` (that we don't consider them `Send`). I'm not 100% sure it's the same problem as the original issue though. We'll see.

**Comment by Dessix** [comment]
For anyone blocked by this in VSCode: Extensions -> rust-analyzer, click the arrow next to "Uninstall", and "Install another version"; `v0.3.986` is the last-known-good before this issue hit.

**Comment by lnicola** [comment]
@Dessix there are instructions in https://github.com/rust-analyzer/rust-analyzer/pull/11780 on how to disable the diagnostic.

**Comment by Veykril** [comment]
Fyi I might've found the problem

**Comment by Veykril** [comment]
#11806 should fix this, the next nightly release (in ~1 hour) ships with that commit

**Comment by lnicola** [comment]
@Veykril did you mean to leave this open?

**Comment by Veykril** [comment]
I wasn't entirely sure whether this has been fixed by my PR or not hence why I didn't close it.

**Comment by Follpvosten** [comment]
As far as I can tell, with `rust-analyzer version: d64f455e8 2022-03-25 nightly`, it still happens, both for async_trait and for the async-graphql attribute macros.

It is fixed for my Rocket project however, where it previously fired on some async blocks.

**Comment by Veykril** [comment]
Ye can confirm the initial bug is fixed, and I can also see async_trait firing some diagnostics. I'll take a small look into it and open a new issue for that.
